### PR TITLE
Address coverity + cleanup pairwise logic

### PR DESCRIPTION
**Pull request by ethanglaser**
## Description

---

<details>
<summary>Checklist:</summary>

**Completeness and readability**

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with updates and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/doc/sources/contribute.rst) for details)_.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least a summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance and/or quality metrics have changed or why changes are not expected.
- [ ] I have extended the benchmarking suite and provided a corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

</details>

### CODE REVIEW

File: daal4py/sklearn/metrics/_pairwise.py [hunk 1]
```diff
@@ -132,54 +102,13 @@ def _pairwise_distances(
...
-        X, _ = check_pairwise_arrays(
-            X, Y, precomputed=True, ensure_all_finite=force_all_finite
-        )
-        whom = (
-            "`pairwise_distances`. Precomputed distance "
-            " need to have non-negative values."
-        )
-        check_non_negative(X, whom=whom)
-        return X
-    if metric in PAIRWISE_DISTANCE_FUNCTIONS:
```
Comment (david-cortes-intel) [inline daal4py/sklearn/metrics/_pairwise.py:151]: I think this would disable oneDAL for the cases where a user passes a supported metric from scikit-learn as a function object.

Comment (ethanglaser) [reply daal4py/sklearn/metrics/_pairwise.py:151]: It's impossible to reach anything below line 140 with the logic in this function

Comment (icfaust) [reply daal4py/sklearn/metrics/_pairwise.py:151]: Just so that I understand, wouldn't @david-cortes-intel 's concern be a gap in the current code (as on line 108 we dont go to daal when its a callable)?

Comment (david-cortes-intel) [reply daal4py/sklearn/metrics/_pairwise.py:151]: Yes, looks like it's also a problem in the current main branch. Could nevertheless be fixed here.

Comment (ethanglaser) [reply daal4py/sklearn/metrics/_pairwise.py:151]: Just to clarify, these deletions are copied directly from the sklearn functionality: https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/metrics/pairwise.py#L2454-L2496 and we have [this condition](https://github.com/uxlfoundation/scikit-learn-intelex/pull/3337/changes#diff-ca2695b03ff78155a0e201c2d28c8132982422fcc481f62f0d1ecb597c09cdd3R64) that uses dal if metric is supported. What more would this help with?

Comment (david-cortes-intel) [reply daal4py/sklearn/metrics/_pairwise.py:151]: The second link doesn't show line number, only a whole-file diff.

Comment (ethanglaser) [reply daal4py/sklearn/metrics/_pairwise.py:151]: This: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/daal4py/sklearn/metrics/_pairwise.py#L88 which is unaffected by this PR

Comment (david-cortes-intel) [reply daal4py/sklearn/metrics/_pairwise.py:151]: @ethanglaser Yes, my point is that those conditions could be changed in this same PR using the dict logic that's currently not reachable, so that they'd work when passing recognized functions.

Comment (ethanglaser) [reply daal4py/sklearn/metrics/_pairwise.py:151]: The scope of the dict doesn't really overlap with the scope of our dal ready metrics though - we have cosine and correlation, the sklearn PAIRWISE_DISTANCE_FUNCTION dict does have cosine but does not have correlation, and has a bunch that we do not support. Additionally, the dict's values are 2-D matrix functions like cosine_distances(X, Y), but a callable metric is invoked per row pair as a scalar (out[i, j] = metric(x, y)), so a user can't pass those functions in the first place — a reverse lookup would never match.

Comment (david-cortes-intel) [reply daal4py/sklearn/metrics/_pairwise.py:151]: Looks like scikit-learn indeed doesn't work like their docs describe.

**Review by david-cortes-intel — APPROVED** [review]
